src/seat/touch.rs

```diff
@@ -0,0 +1,184 @@
+    fn up(
+        &mut self,
+        conn: &mut ConnectionHandle,
+        qh: &QueueHandle<Self>,
+        touch: &WlTouch,
+        serial: u32,
+        time: u32,
+        id: i32,
+    );
+
+    /// Touch point motion.
+    #[allow(clippy::too_many_arguments)]
+    fn motion(
+        &mut self,
+        conn: &mut ConnectionHandle,
+        qh: &QueueHandle<Self>,
+        touch: &WlTouch,
+        time: u32,
+        id: i32,
+        x: f64,
+        y: f64,
+    );
+
+    /// Touch point shape change.
+    ///
+    /// The shape of a touch point is approximated by an ellipse through the major and minor axis
+    /// length. Major always represents the larger of the two axis and is orthogonal to minor.
+    ///
+    /// The dimensions are specified in surface-local coordinates and the locations reported by
+    /// other events always report the center of the ellipse.
+    fn shape(
+        &mut self,
+        conn: &mut ConnectionHandle,
+        qh: &QueueHandle<Self>,
+        touch: &WlTouch,
+        id: i32,
+        major: f64,
+        minor: f64,
+    );
+
+    /// Touch point shape orientation.
+    ///
+    /// The orientation describes the clockwise angle of a touch point's major axis to the positive
+    /// surface y-axis and is normalized to the -180° to +180° range.
+    fn orientation(
+        &mut self,
+        conn: &mut ConnectionHandle,
+        qh: &QueueHandle<Self>,
+        touch: &WlTouch,
+        id: i32,
+        orientation: f64,
+    );
+
+    /// Cancel active touch sequence.
+    ///
+    /// This indicates that the compositor has cancelled the active touch sequence, for example due
+    /// to detection of a touch gesture.
+    fn cancel(&mut self, conn: &mut ConnectionHandle, qh: &QueueHandle<Self>, touch: &WlTouch);
+}
+
+impl DelegateDispatchBase<WlTouch> for SeatState {
+    type UserData = TouchData;
+}
+
+impl<D> DelegateDispatch<WlTouch, D> for SeatState
+where
+    D: Dispatch<WlTouch, UserData = Self::UserData> + TouchHandler,
+{
+    fn event(
+        data: &mut D,
+        touch: &WlTouch,
+        event: TouchEvent,
+        udata: &Self::UserData,
+        conn: &mut ConnectionHandle,
+        qh: &QueueHandle<D>,
+    ) {
+        match event {
+            // Buffer events until frame is received.
+            TouchEvent::Down { .. }
+            | TouchEvent::Up { .. }
+            | TouchEvent::Motion { .. }
+            | TouchEvent::Shape { .. }
+            | TouchEvent::Orientation { .. } => {
+                let mut guard = udata.inner.lock().unwrap();
+                guard.events.push(event);
+            }
+            // Process all buffered events.
+            TouchEvent::Frame => {
+                let mut guard = udata.inner.lock().unwrap();
+                for event in guard.events.drain(..) {
+                    process_framed_event(data, touch, conn, qh, event);
+                }
+            }
+            TouchEvent::Cancel => data.cancel(conn, qh, touch),
```

Comment:
You mean on cancel? I'm not sure to be honest.

Generally I don't think the cancel event is tied to frame events since it's created by the compositor. What I would assume is that a cancel is generally only transmitted after all relevant events have been framed, since only on frame should they actually be processed to be recognized as a gesture?

Like the frame isn't meant to be used to buffer touch input that isn't detected as gesture yet, but more as a debounce mechanism? Though I suppose it probably shouldn't *hurt* to just clear it when we assume all good-behaving compositors will never call cancel with a pending frame event?

Reply:
> Though I suppose it probably shouldn't hurt to just clear it when we assume all good-behaving compositors will never call cancel with a pending frame event?

Yeah, that's mostly what I have in mind